src/ModelFactory.cpp

```diff
@@ -43,43 +50,101 @@
 
 namespace PacBio {
 namespace Consensus {
+namespace {
 
 using ChemistryNotFound = PacBio::Exception::ChemistryNotFound;
 
-ModelCreator::ModelCreator(const std::set<std::string>& names)
+size_t Count(const std::string& str, const std::string& delim)
 {
-    for (const std::string& name : names)
-        if (!ModelFactory::Register(name, this))
-            throw std::runtime_error("duplicate model inserted into factory!");
+    size_t count = 0;
+    size_t start = 0;
+
+    while ((start = str.find(delim, start)) != std::string::npos) {
+        ++count;
+        start += delim.length();
+    }
+
+    return count;
+}
+
 }
 
 std::unique_ptr<ModelConfig> ModelFactory::Create(const std::string& name, const SNR& snr)
 {
-    const auto it = CreatorTable().find(name);
+    boost::optional<std::string> model(boost::none);
+
+    if (!(model = ModelOverride()))
+        if (!(model = Resolve(name)))
+            throw ChemistryNotFound(name);
 
-    if (it == CreatorTable().end()) throw ChemistryNotFound(name);
+    const auto& tbl = CreatorTable();
+    const auto it = tbl.find(*model);
+
+    if (it == tbl.end()) throw ChemistryNotFound(name);
 
     return it->second->Create(snr);
 }
 
-bool ModelFactory::Register(const std::string& name, ModelCreator* const ctor)
+bool ModelFactory::Register(const std::string& name, std::unique_ptr<ModelCreator>&& ctor)
+{
+    return CreatorTable()
+        .insert(std::make_pair(name, std::forward<std::unique_ptr<ModelCreator>>(ctor)))
+        .second;
+}
+
+boost::optional<std::string> ModelFactory::Resolve(const std::string& name)
```

Comment:
In Ocaml, this monad is the right thing.

In C++, I hate boost because it compiles soooooo slooooooowly. I'm not sure about the cost of boost::optional, but why bother when there are 2 very simple alternatives:

1. Raise an exception on failure, or
2. Return the result as an output parameter, and return an error-code.